### PR TITLE
`Development`: Fix issues with LTI 1.3 deep linking service for openedX

### PR DESCRIPTION
**Pull request by basak-akan**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

This PR fixes current deep linking service implementation to be able used by openedX configurations as well.

### Description

- Combined the 'auth-callback' and 'deep-link' endpoints into a single controller method. This streamlines application routing, ensuring consistent handling of related LTI 1.3 processes. 
- Corrected the deep linking response content items JSON to make it more universally parsable across various scenarios.

### Steps for Testing

Prerequisites:
- Please deploy to TS5 where lti profile is enable and already prepared for easy testing

1. Log in to openedX Studio https://studio.openedx.ase.cit.tum.de/home/ (Studio is the edX tool you use to build your courses) with one of the instructors (i.e. artemis_test_user_11)
2. Click on "Introduction" section
3. Navigate to first "Unit" subsection
4. Click on the link at the bottom part of the LTI Consumer Block "Deep Linking Launch - Configure tool"
5. Verify you will be directed to Artemis login page where you need to login using same instructor credentials.
6. Select the couse and verify you are navigated into selecting exercises page.
7. Select one of the exercises listed.
8. Verify you see deep linking successfull and close the frame message on the page.
9. Close the tab and go back to previous tab refresh the page and verify you see selected exercise link and title as json object on the bottom side of the lti consumer block.  

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [edit-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/edit-lti-configuration.component.ts.html) | 85% | ✅ |
| [lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/lti-configuration.component.ts.html) | 94.73% | ✅ |
| [lti-configuration.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/lti-configuration.service.ts.html) | 88.88% | ✅ |
| [edit-course-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-lti-configuration/edit-course-lti-configuration.component.ts.html) | 82.5% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [Lti13DeepLinkingResponse.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Lti13DeepLinkingResponse.html) | 82% | ✅ |
| [Lti13LaunchRequest.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Lti13LaunchRequest.html) | 97% | ✅ |
| [LtiResourceLaunch.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/LtiResourceLaunch.html) | 76% | ✅ |
| [LtiPlatformConfigurationRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/LtiPlatformConfigurationRepository.html) | 76% | ✅ |
| [Lti13LaunchFilter.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.lti/Lti13LaunchFilter.html) | 100% | ✅ |
| [Lti13Service.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/Lti13Service.html) | 88% | ✅ |
| [LtiDeepLinkingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/LtiDeepLinkingService.html) | 97% | ✅ |
| [AdminLtiConfigurationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminLtiConfigurationResource.html) | 86% | ✅ |
| [OnlineCourseDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/OnlineCourseDTO.html) | 100% | ✅ |
| [PublicLtiResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6129/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.open/PublicLtiResource.html) | 97% | ✅ |

### Screenshots

image

## Summary by CodeRabbit

- **New Features**
	- Enhanced LTI configuration management in the admin panel with updated URL paths and constants.
- **Refactor**
	- Updated deep linking configuration and redirection paths to enhance LTI 1.3 support.
- **Tests**
	- Adjusted tests to reflect changes in deep linking URL paths and content item representation.

### CODE REVIEW

**Review by muradium — DISMISSED** [review]
Tested with TS5, works as expected

**Review by egekurt123 — APPROVED** [review]
Manually tested on TS5, works as expected

**Comment by beyzaaltuntas** [comment]
Manually tested on TS5, works as expected

**Review by beyzaaltuntas — APPROVED** [review]
Manually tested on TS5, works as expected

**Review by egekurt123 — APPROVED** [review]
Reapprove after solving merge conflicts

**Review by beyzaaltuntas — DISMISSED** [review]

**Review by muradium — APPROVED** [review]

**Review by rstief — DISMISSED** [review]
Code 👍

File: src/main/webapp/app/admin/lti-configuration/lti-configuration.component.ts [hunk 1]
```diff
@@ -62,7 +62,7 @@ export class LtiConfigurationComponent implements OnInit {
      * Gets the deep linking url
      */
     getDeepLinkingUrl(): string {
-        return `${location.origin}/lti/deep-linking`; // Needs to match url in CustomLti13Configurer
+        return `${location.origin}/api/public/lti13/deep-link`; // Needs to match url in CustomLti13Configurer
```
Comment (loreanvictor) [inline src/main/webapp/app/admin/lti-configuration/lti-configuration.component.ts:65]: I'd recommend refactoring these hard-coded values into (potentially static) variables that other parts of the code read from and compose URLs based on (similar to server-side code). `'lti13/...'` is a recurring constant in this particular component, and these implicit couplings are prone to errors (someone might forget updating one value while updating another, for example).

File: src/main/java/de/tum/in/www1/artemis/security/lti/Lti13LaunchFilter.java [hunk 2]
```diff
@@ -66,6 +67,10 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
             try {
                 // here we need to check if this is a deep-linking request or a launch request
                 if ("LtiDeepLinkingRequest".equals(ltiIdToken.getClaim(Claims.MESSAGE_TYPE))) {
+                    // Manually setting the deep linking path is required due to Moodle and edX's inconsistent deep linking implementation.
+                    // Unlike standard GET request-based methods, these platforms do not guarantee a uniform approach, necessitating
+                    // manual configuration to ensure reliable navigation and resource access compatibility.
+                    targetLink = CustomLti13Configurer.LTI13_DEEPLINKING_REDIRECT_PATH;
```
Comment (pzdr7) [inline src/main/java/de/tum/in/www1/artemis/security/lti/Lti13LaunchFilter.java:73]: I think it's good to use a constant here - perhaps the string `"LtiDeepLinkingRequest"` above should be a constant as well (it's used a few times).

Comment (basak-akan) [reply src/main/java/de/tum/in/www1/artemis/security/lti/Lti13LaunchFilter.java]: Thanks for the review @pzdr7, "LtiDeepLinkingRequest" is now a constant

File: src/main/java/de/tum/in/www1/artemis/web/rest/open/PublicLtiResource.java [hunk 3]
```diff
@@ -37,12 +37,16 @@ public class PublicLtiResource {
...
+     *
+     * Consolidates handling for both 'auth-callback' and 'deep-link' endpoints to simplify client interactions.
+     * This approach ensures consistent processing and user experience for authentication and deep linking flows.
      *
      * @param request  HTTP request
      * @param response HTTP response
      * @throws IOException If an input or output exception occurs
      */
-    @PostMapping("lti13/auth-callback")
+    @PostMapping({ "lti13/auth-callback", "lti13/deep-link" })
```
[the remark this answers is not in the document]
Comment (basak-akan) [reply src/main/java/de/tum/in/www1/artemis/web/rest/open/PublicLtiResource.java]: Thanks! @loreanvictor they are now retrieved from `CustomLti13Configurer` 👍

File: src/main/java/de/tum/in/www1/artemis/config/lti/CustomLti13Configurer.java [hunk 4]
```diff
@@ -44,7 +44,10 @@ public class CustomLti13Configurer extends Lti13Configurer {
     public static final String LTI13_LOGIN_REDIRECT_PROXY_PATH = LTI13_BASE_PATH + "/auth-callback";
 
     /** Path for LTI 1.3 deep linking. */
-    public static final String LTI13_DEEPLINKING_PATH = "/lti/deep-linking/";
+    public static final String LTI13_DEEPLINKING_PATH = LTI13_BASE_PATH + "/deep-link";
+
+    /** Path for LTI 1.3 deep linking redirect. */
+    public static final String LTI13_DEEPLINKING_REDIRECT_PATH = "/lti/deep-linking/";
```
[the remark this answers is not in the document]
Comment (basak-akan) [reply src/main/java/de/tum/in/www1/artemis/config/lti/CustomLti13Configurer.java]: You're correct, and I appreciate your attention to detail. I've updated the URL names to better reflect their intended uses.

**Review by pzdr7 — DISMISSED** [review]
Code

**Review by loreanvictor — DISMISSED** [review]
code lgtm

**Review by muradium — APPROVED** [review]

**Review by egekurt123 — APPROVED** [review]
Reapprove

**Pull request merged**